infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java

```diff
@@ -210,23 +255,123 @@ private void addMachineVolumes(
+
+      container.getVolumeMounts().add(newVolumeMount(podVolume.getName(), volumePath, subPath));
     }
   }
 
-  private void addVolumeIfNeeded(String pvcName, PodSpec podSpec) {
-    if (podSpec.getVolumes().stream().noneMatch(volume -> volume.getName().equals(pvcName))) {
-      podSpec.getVolumes().add(newVolume(pvcName, pvcName));
+  private void replacePodsVolumesWithCommon(Map<String, PodData> pods, String commonPVCName) {
+    for (PodData pod : pods.values()) {
+      Set<String> pvcVolumes = new HashSet<>();
+
+      Iterator<io.fabric8.kubernetes.api.model.Volume> volumeIterator =
+          pod.getSpec().getVolumes().iterator();
+      while (volumeIterator.hasNext()) {
+        io.fabric8.kubernetes.api.model.Volume volume = volumeIterator.next();
+        if (volume.getPersistentVolumeClaim() != null) {
+          pvcVolumes.add(volume.getName());
+          volumeIterator.remove();
+        }
+      }
+
+      pod.getSpec()
+          .getVolumes()
+          .add(
+              new VolumeBuilder()
+                  .withName(commonPVCName)
+                  .withNewPersistentVolumeClaim()
+                  .withClaimName(commonPVCName)
+                  .endPersistentVolumeClaim()
+                  .build());
+
+      Stream.concat(
+              pod.getSpec().getContainers().stream(), pod.getSpec().getInitContainers().stream())
+          .flatMap(c -> c.getVolumeMounts().stream())
+          .forEach(
+              volumeMount -> {
+                if (pvcVolumes.contains(volumeMount.getName())) {
+                  volumeMount.setName(commonPVCName);
+                }
+              });
     }
   }
 
-  /** Get sub-path that holds all the volumes of a particular workspace */
-  private String getWorkspaceSubPath(String workspaceId) {
-    return workspaceId;
+  private void prefixVolumeMountsSubpaths(
+      KubernetesEnvironment k8sEnv, String workspaceId, Set<String> subPaths) {
+    for (PodData pod : k8sEnv.getPodsData().values()) {
+      Map<String, String> volumeToClaimName = new HashMap<>();
+      for (io.fabric8.kubernetes.api.model.Volume volume : pod.getSpec().getVolumes()) {
+        if (volume.getPersistentVolumeClaim() == null) {
+          continue;
+        }
+        volumeToClaimName.put(volume.getName(), volume.getPersistentVolumeClaim().getClaimName());
+      }
+
+      if (volumeToClaimName.isEmpty()) {
+        // Pod does not have any volume that references PVC
+        continue;
+      }
+
+      Stream.concat(
+              pod.getSpec().getContainers().stream(), pod.getSpec().getInitContainers().stream())
+          .forEach(
+              c -> {
+                for (VolumeMount volumeMount : c.getVolumeMounts()) {
+                  String pvcName = volumeToClaimName.get(volumeMount.getName());
+                  if (pvcName == null) {
+                    k8sEnv
+                        .getWarnings()
+                        .add(
+                            new WarningImpl(
```

Comment:
I've refactored KubernetesEnvValidator and added validation of `PVC`s and `Pod#Volumes`, `Container#VolumesMounts` there. So, skipping such volume mounts without warnings since it should not happen but check on null still exist to prevent NPE.